File: sdk/chocobase-rb/lib/chocobase.rb
require "net/http"
require "json"
require "uri"

require_relative "chocobase/auth"
require_relative "chocobase/postgrest"
require_relative "chocobase/storage"
require_relative "chocobase/functions"

module ChocoBase
  class Client
    attr_reader :url, :api_key, :headers, :auth, :storage, :functions

    def initialize(url, api_key, custom_headers: {})
      @url = url.chomp("/")
      @api_key = api_key
      @headers = {
        "apikey" => api_key,
        "Authorization" => "Bearer #{api_key}",
        "Content-Type" => "application/json"
      }.merge(custom_headers)

      @auth = Auth.new(@url, @headers)
      @storage = Storage.new(@url, @headers)
      @functions = Functions.new(@url, @headers)
    end

    def from(table)
      Postgrest.new(@url, table, @headers)
    end
  end

  def self.create_client(url, api_key, custom_headers: {})
    Client.new(url, api_key, custom_headers: custom_headers)
  end
end

File: sdk/chocobase-jl/src/ChocoBase.jl
module ChocoBase

using HTTP
using JSON3

export ChocoClient, from, sign_up, sign_in, select, eq, limit, execute, create_signed_url, invoke

struct ChocoClient
    url::String
    api_key::String
    headers::Dict{String, String}

    function ChocoClient(url::String, api_key::String, custom_headers::Dict{String, String}=Dict{String, String}())
        clean_url = rstrip(url, '/')
        base_headers = Dict(
            "apikey" => api_key,
            "Authorization" => "Bearer " * api_key,
            "Content-Type" => "application/json"
        )
        merge!(base_headers, custom_headers)
        new(clean_url, api_key, base_headers)
    end
end

include("Auth.jl")
include("Postgrest.jl")
include("Storage.jl")
include("Functions.jl")

end # module
